src/main/java/com/hedera/mirror/downloader/Downloader.java

```
@@ -270,8 +272,113 @@ protected boolean moveFile(File sourceFile, File destinationFile) {
+                                applicationStatusRepository.updateStatusValue(getLastValidDownloadedFileHashKey(), hash);
+                            }
+                            applicationStatusRepository.updateStatusValue(getLastValidDownloadedFileKey(), destination.getName());
+                            valid = true;
+                            break;
+                        }
+                    }
+                } else if (file != null) {
+                    log.warn("Hash doesn't match the hash contained in valid signature file. Will try to download a file with same timestamp from other nodes and check the Hash: {}", file);
+                }
+            }
+
+            if (!valid) {
+                log.error("File could not be verified by at least 2/3 of nodes: {}", sigFileName);
+            }
+        }
+    }
+
+
+    /**
+     * Verifies that prevFileHash in given {@code file} matches that in application repository.
+     * @throws Exception
+     */
+    private boolean verifyHashChain(File file) {
+        String filePath = file.getAbsolutePath();
+        String lastValidFileHash = applicationStatusRepository.findByStatusCode(getLastValidDownloadedFileHashKey()); // todo
+        String bypassMismatch = applicationStatusRepository.findByStatusCode(getBypassHashKey());
+        String prevFileHash = getPrevFileHash(filePath);
+
+        if (prevFileHash == null) {
+            log.warn("Doesn't contain valid previous file hash: {}", filePath);
+            return false;
+        }
+
+        if (StringUtils.isBlank(lastValidFileHash) || lastValidFileHash.equals(prevFileHash) ||
+                Utility.hashIsEmpty(prevFileHash) || bypassMismatch.compareTo(file.getName()) > 0) {
+            return true;
+        }
+
+        log.warn("File Hash Mismatch with previous: {}, expected {}, got {}", file.getName(), lastValidFileHash, prevFileHash);
+        return false;
+    }
+
+    protected Pair<Boolean, File> downloadFile(File sigFile) {
```

Let's change this to just return `File` and remove the `Pair`. Remove the try/catch and let it bubble up to `verifySigsAndDownloadDataFiles`. Return null if `pendingDownload.isDownloadSuccessful()` is false.

---

Had this change in followup PR, happy to pull it in.